keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,1100 @@
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `requiredIndexes` to 64Ki.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +option
+	SuccessfulPolicy *SuccessfulPolicy
+}
+
+// SuccessfulPolicy describes how succeeded pods influence the conditions.
+type SuccessfulPolicy struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +option
+	RequiredIndexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up. 
+	// 
+	// +option
+	MinSucceededIndexes intstr.IntOrString
+}
+```
+
+Moreover, we validate the following constraints:
+- exactly one of the fields `requiredIndexes` and `minSucceededIndexes` is specified
+  for a requirement.
+- whether the specified indexes in the `requiredIndexes` and
+  the number of indexes in the `minSucceededIndexes` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `requiredIndexes` is under 64Ki. 
+
+### Evaluation
+We evaluate the SuccessfulPolicy after all indexes have been finished.
```

Comment:
> I don't think this is what the stories above suggest.
They suggest that the policy is satisfied as soon as the desired pods succeed. The workers might not even have the ability to terminate on their own, like MPI workers.

It makes sense.

> what happens with the running Pods when the success criteria is reached? Are they terminated?

I think expected behavior differs by use cases and policies (`indexes` vs `count`). 

if using `indexes`, some of the indexes are leaders, and all leaders succeeded means the whole of batch job succeeded. So we would propose that we terminate the running pods when batch jobs meet the successPolicy. 

If using `count`, expected behaviors are different by use cases. For example, let's imagine the situation that all indexes play a worker role:

1. Start batch job with `completions=10` and `count=5`.
2. the batch job is scaled in `completions=8` by autoscaler.
3. We would expect that all 8 indexes succeeded and don't expect that the job controller terminates running pods even if the batch job meets successPolicy (`count=5`).

Also, we can imagine the situation that all indexes play leader role (HA leaders):

1. Start batch job with `completions=3` and `count=1`.
2. One of the leaders succeeded.
3. We would expect that one of the indexes succeeded and the job controller terminates running pods once the batch job meets successPolicy (`count=1`).

So adding an action field like `podFilaurePolicy` would be good. @alculquicondor wdyt?

> Do they count towards failed or succeeded or none?

I would propose we count them towards success or none (unknown). Because if we count them towards failed, we need to consider backOffLimit, and it is able to produce potential bugs.

Reply:
> If using `count`, expected behaviors are different by use cases. For example, let's imagine the situation that all indexes play a worker role:
> 
> 1. Start batch job with `completions=10` and `count=5`.
> 2. the batch job is scaled in `completions=8` by autoscaler.
> 3. We would expect that all 8 indexes succeeded and don't expect that the job controller terminates running pods even if the batch job meets successPolicy (`count=5`).

So all pods should terminate, but if 5 have phase=Success, then we still mark the Job as succeeded, even if the rest of the indexes fail?
Do you have a more specific use case that would work like this?

> 
> Also, we can imagine the situation that all indexes play leader role (HA leaders):
> 
> 1. Start batch job with `completions=3` and `count=1`.
> 2. One of the leaders succeeded.
> 3. We would expect that one of the indexes succeeded and the job controller terminates running pods once the batch job meets successPolicy (`count=1`).

I can think of a case that works like this, but where there are no leaders:
Imagine you are trying to minimize the value of a multi-parameter function and you want to find a solution for a target threshold X. You start an indexed job where each pod implements gradient descent has a different starting point for the parameters. Once each index reaches the threshold, it finishes, if not, it continues searching. Then, for the entire job you can set count=1, so that once one solution is found, we no longer continue searching for other solutions.

> 
> So adding an action field like `podFilaurePolicy` would be good. wdyt?

Maybe? I would like to see a use case for the first scenario.

> 
> > Do they count towards failed or succeeded or none?
> 
> I would propose we count them towards `succeeded` or `none` ( or `unknown`). Because if we count them towards `failed`, we need to consider backOffLimit, and it is able to produce potential bugs.

Reply:
> Do you have a more specific use case that would work like this?
Perhaps some simulation / data generation that is known to error / fail sometimes (for any myriad of reasons), where you want at least N samples, hope for N+M, but would be happy with a value in-between that.

Reply:
I see.... That makes sense to me. Then it sounds like there should be a field controlling whether to terminate the running Pods.

`whenTargetAchieved: Continue | TerminateRunning`

Reply:
Done.